src/main/java/org/opensearch/ad/ratelimit/CheckpointWriteWorker.java

```diff
@@ -0,0 +1,299 @@
+    ) {
+        super(
+            "checkpoint-write",
+            heapSizeInBytes,
+            singleRequestSizeInBytes,
+            maxHeapPercentForQueueSetting,
+            clusterService,
+            random,
+            adCircuitBreakerService,
+            threadPool,
+            settings,
+            maxQueuedTaskRatio,
+            clock,
+            mediumSegmentPruneRatio,
+            lowSegmentPruneRatio,
+            maintenanceFreqConstant,
+            clientUtil,
+            CHECKPOINT_WRITE_QUEUE_CONCURRENCY,
+            executionTtl,
+            CHECKPOINT_WRITE_QUEUE_BATCH_SIZE,
+            stateTtl,
+            stateManager
+        );
+        this.indexUtil = indexUtil;
+        this.checkpoint = checkpoint;
+        this.indexName = indexName;
+        this.checkpointInterval = checkpointInterval;
+    }
+
+    @Override
+    protected void executeBatchRequest(BulkRequest request, ActionListener<BulkResponse> listener) {
+        if (indexUtil.doesCheckpointIndexExist()) {
+            clientUtil.<BulkRequest, BulkResponse>execute(BulkAction.INSTANCE, request, listener);
+        } else {
+            indexUtil.initCheckpointIndex(ActionListener.wrap(initResponse -> {
+                if (initResponse.isAcknowledged()) {
+                    clientUtil.<BulkRequest, BulkResponse>execute(BulkAction.INSTANCE, request, listener);
+                } else {
+                    throw new RuntimeException("Creating checkpoint with mappings call not acknowledged.");
+                }
+            }, exception -> {
+                if (ExceptionsHelper.unwrapCause(exception) instanceof ResourceAlreadyExistsException) {
+                    // It is possible the index has been created while we sending the create request
+                    clientUtil.<BulkRequest, BulkResponse>execute(BulkAction.INSTANCE, request, listener);
+                } else {
+                    LOG.error(String.format(Locale.ROOT, "Unexpected error creating checkpoint index"), exception);
+                    listener.onFailure(exception);
+                }
+            }));
+        }
+    }
+
+    @Override
+    protected BulkRequest toBatchRequest(List<CheckpointWriteRequest> toProcess) {
+        final BulkRequest bulkRequest = new BulkRequest();
+        for (CheckpointWriteRequest request : toProcess) {
+            bulkRequest.add(request.getIndexRequest());
+        }
+        return bulkRequest;
+    }
+
+    @Override
+    protected ActionListener<BulkResponse> getResponseListener(List<CheckpointWriteRequest> toProcess, BulkRequest batchRequest) {
+        return ActionListener.wrap(response -> {
+            for (BulkItemResponse r : response.getItems()) {
+                if (r.getFailureMessage() != null) {
+                    // maybe indicating a bug
+                    // don't retry failed requests since checkpoints are too large (250KB+)
+                    // Later maintenance window or cold start will retry saving
```

Comment:
Why not throw exception here?

Reply:
If throwing exceptions, I don't know how to handle it.  If we throw it to Opensearch, we risk bring the cluster down.  Log an exception so that we may be able to debug.

Reply:
Ah I see thanks